nemo/lightning/pytorch/strategies.py

```diff
@@ -167,6 +164,21 @@ def connect(self, model: pl.LightningModule) -> None:
             config.sequence_parallel = self.sequence_parallel
             self._mcore_config = config
 
+        has_optim = getattr(model, "optim", None)
+        if has_optim:
+            opt_config = getattr(model.optim, "config", None)
+            if isinstance(opt_config, OptimizerConfig):
+                mcore_opt_config: OptimizerConfig = cast(OptimizerConfig, opt_config)
+                if not self.ddp_config:
+                    raise ValueError("PyTorch DDP is not enabled for mcore optimizer")
+                ddp_config = cast(DistributedDataParallelConfig, self.ddp_config)
+
+                if mcore_opt_config.use_distributed_optimizer != ddp_config.use_distributed_optimizer:
```

Comment:
does that mean we have two entry points for a single option?

Reply:
Yeah that's how mcore does it currently, I don't like that design that much. Hopefully we can improve mcore's design around this in a near future release.

Reply:
Hmm, how about the following: remove the patching here, convert it an assertion & make the ddp_config's use_distributed_optimizer value dependent on the mcore_opt_config? would that be possible>

Reply:
Yeah maybe we should remove [this](https://github.com/NVIDIA/NeMo/blob/main/nemo/lightning/pytorch/strategies.py#L135-L136) from the constructor and only initialize it here when we detect the mcore optimizer? Then the default code-path is easier to follow, when the user specifies a custom ddp_config we could as you said throw an exception when the values don't match. You think that makes sense?